### PR TITLE
  [Diffusion] [Model] Add MOVA video-audio generation from OpenMOSS

### LINKED ISSUE #2079: [New Model]: MOVA from OpenMOSS
### The model to consider.

https://github.com/OpenMOSS/MOVA

### The closest model vllm-omni already supports.

Helios/DreamID-Omni

### What's your difficulty of supporting the model you want?

_No response_

### Use case and motivation

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by reidliu41**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose

  Add base support for [MOVA-360p](https://github.com/OpenMOSS/MOVA) (OpenMOSS) for offline I2VA
  (image-to-video-audio) inference. Closes #2079.

  MOVA generates synchronized video and audio in a single pass using a dual-tower architecture: a 14B video DiT, a
   1.3B audio DiT, and a cross-modal bridge with Aligned RoPE for temporal synchronization.

## Test Plan

**1. Download model and test assets:**

  ```bash
  # Download MOVA-360p model 
  huggingface-cli download OpenMOSS-Team/MOVA-360p --local-dir ~/test-models/MOVA-360p

  # Clone upstream repo for test reference image
  git clone --depth 1 https://github.com/OpenMOSS/MOVA /tmp/MOVA-repo
  ```

  **2. Run inference:**

  ```bash
  python examples/offline_inference/mova/end2end.py \
      --model ~/test-models/MOVA-360p \
      --prompt "a person talking and waving" \
      --ref-path /tmp/MOVA-repo/assets/single_person.jpg \
      --height 352 --width 640 \
      --num-inference-steps 5 \
      --seed 42 \
      --output mova_test.mp4
  ```

## Test Result

  | Metric | Value |
  |---|---|
  | Model loading | All components, zero missing keys |
  | Denoising | 5/5 steps completed |
  | Boundary switch | video_dit → video_dit_2 |
  | Video output | 193 frames, 352×640 |
  | Audio output | 8.06s, 48kHz, mono |
  | MP4 muxing | successful |

https://github.com/user-attachments/assets/77e5b7e9-8a0e-4fa0-a89d-1db2797c546e

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by 0-693** [comment]
@reidliu41 Hi, I'm from the MOVA team. We ran your code and here are our test results and some suggestions.
The first video was tested with a more precise prompt + seed=67 at step 5:
> A man in a blue blazer and glasses speaks in a formal indoor setting, framed by wooden furniture and a filled bookshelf. Quiet room acoustics underscore his measured tone as he delivers his remarks. At one point, he says, "I would also believe that this advance in AI recently was quite predictable."
We tested both step 5 and step 50 across multiple test samples. The third video performed noticeably better at step 50.


Currently, without integrating a prompt rewriting system, our observations are:
1. We recommend using complete and detailed prompts for testing. Short or vague prompts tend to produce unstable and often unusable outputs.
2. We recommend step 50, especially for complex scenes. At step 5, the video often fails to render properly — it's not just a matter of minor quality degradation, but the output being fundamentally unusable.

One additional issue to flag: in our testing, the current vLLM inference PSNR avg score is only ~27dB, which is below the acceptable threshold of 30dB+. This may require investigating inference precision or quantization-related configurations.
Hope this feedback is helpful.

https://github.com/user-attachments/assets/0323060a-b6c9-4127-8716-041c492ea438


https://github.com/user-attachments/assets/613b3795-3276-4cae-95c9-685e69c865ce


https://github.com/user-attachments/assets/765fb97e-5cb7-42e3-badd-9db6c19551b0

File: vllm_omni/diffusion/models/mova/scheduling_mova.py [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+    def set_timesteps(self, *args, **kwargs) -> None:
+        super().set_timesteps(*args, **kwargs)
+        self._refresh_pair_cache()
+
+    def timestep_to_sigma(self, timestep: torch.Tensor | float) -> torch.Tensor:
+        """Return the sigma for a given timestep via nearest-neighbor lookup."""
+        t_value = float(timestep)
+        t_cpu = torch.tensor(t_value)
+        idx = torch.argmin((self.train_timesteps - t_cpu).abs())
+        return self.train_sigmas[idx]
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/mova/scheduling_mova.py:363]: The bot comment below has a point. `step_from_to()` calls `timestep_to_sigma()` which does nearest-neighbor lookup in `self.train_timesteps`/`self.train_sigmas` — cached from the base schedule *before* `set_pair_postprocess_by_name("dual_sigma_shift")` runs. So when the pipeline passes paired timesteps with different visual/audio shifts, the sigma lookup still uses the stale base schedule.

This means the `(sigma_to - sigma_from)` delta is wrong whenever `visual_shift` or `audio_shift` differ from the initial scheduler shift. For the default case (both == 5.0 == initial shift) it happens to work, but it'll silently produce wrong results for other configs.

File: pyproject.toml [hunk 2]
```diff
@@ -213,6 +213,7 @@ extend-ignore-identifiers-re = [
     ".*nin.*",
     ".*[Oo]no_[Aa]nna.*",
     ".*cann.*",
+    ".*seperated.*",
```
Comment (lishunyang12) [inline pyproject.toml]: Why not fix the typo (`seperated_timestep` → `separated_timestep`) in the model configs/constructors instead of adding it to the spellcheck ignore list?

File: vllm_omni/diffusion/models/mova/mova_audio_transformer.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    end: int = 16384,
+    theta: float = 10000.0,
+    base_tps: float = 4.0,
+    target_tps: float = 44100.0 / 2048.0,
+) -> torch.Tensor:
+    """
+    Audio RoPE for 'oobleck' VAE type.
+    Applies time-scale factor and sets H/W dims to ones (no spatial encoding).
+    """
+    s = base_tps / target_tps
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/mova/mova_audio_transformer.py:39]: Dead code: `freqs_t` from `precompute_freqs_cis(t_dim, end, theta)` is immediately overwritten below. Drop the first call.

```suggestion
    s = base_tps / target_tps
    t_dim = dim - 2 * (dim // 3)
    h_dim = dim // 3
    w_dim = dim // 3

    # Scale temporal frequencies
    t = torch.arange(end, dtype=torch.float32) * s
    freqs_scaled = 1.0 / (theta ** (torch.arange(0, t_dim, 2, dtype=torch.float32) / t_dim))
    freqs_t = torch.polar(
        torch.ones(end, len(freqs_scaled)),
        torch.outer(t, freqs_scaled),
    )
```

**Review by lishunyang12 — COMMENTED** [review]
left a few comments

File: vllm_omni/entrypoints/async_omni_diffusion.py [hunk 4]
```diff
@@ -109,7 +109,17 @@ def __init__(
...
 
                 tf_config_dict = get_hf_file_to_dict("transformer/config.json", od_config.model)
-                od_config.tf_model_config = TransformerConfig.from_dict(tf_config_dict)
+                if tf_config_dict is not None:
+                    od_config.tf_model_config = TransformerConfig.from_dict(tf_config_dict)
+                elif od_config.model_class_name in ("MOVA", "DreamIDOmniPipeline"):
+                    # Multi-component models have no single transformer/ subfolder
+                    # (e.g. MOVA: video_dit/, audio_dit/, dual_tower_bridge/)
+                    od_config.tf_model_config = TransformerConfig()
+                else:
```
Comment (lishunyang12) [inline vllm_omni/entrypoints/async_omni_diffusion.py:129]: Hard-coding model class names in a fallback allowlist doesn't scale. Next multi-component model will need to touch this file too. Could this check for the existence of `transformer/config.json` and fall back to a default `TransformerConfig()` generically, or add a flag in the registry entry?

**Comment by princepride** [comment]
@reidliu41 Hi, how is the progress?

**Comment by reidliu41** [comment]
I investigated the PSNR concern in detail.

  The current conclusion is that the remaining PSNR gap should not be treated as a simple vLLM-vs-upstream core parity issue. On the direct single-GPU inference path, vllm-omni MOVA is already reasonably close to the upstream MOVA pipeline.
  The large gap appears when comparing against workflow outputs produced by the upstream SGLang serving path, which uses multi-GPU context parallelism.

  At the moment, MOVA in vllm-omni only supports single-GPU direct inference. It does not yet implement the upstream
  context-parallel execution path for MOVA. Because of that, PSNR comparisons against workflow references generated with context parallelism are not an apples-to-apples parity check.

 For this PR, the actionable review comments have been addressed, and the single-GPU MOVA path is usable. Full workflow-level parity with upstream SGLang context-parallel serving should be handled as a separate follow-up task.

**Comment by BeatSeat** [comment]
Any progress here? Is the PSNR acceptable now?
